src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -226,54 +224,68 @@ public override TaintedDataAbstractValue VisitObjectCreation(IObjectCreationOper
+                        out taintedTargets))
                     {
-                        ImmutableArray<IArgumentOperation> argumentOperation = (originalOperation as IInvocationOperation).Arguments;
-                        IEnumerable<PointsToAbstractValue> pointsToAnalysisResult = argumentOperation.Select(o => GetPointsToAbstractValue(o.Value));
-                        IEnumerable<ValueContentAbstractValue> valueContentAnalysisResult = argumentOperation.Select(o => GetValueContentAbstractValue(o.Value));
-                        if ((evaluateWithPointsToAnalysis != null && evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResult)))
-                            || (evaluateWithValueContentAnalysis != null && evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResult, valueContentAnalysisResult))))
+                        foreach (string taintedTarget in taintedTargets)
                         {
-                            return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                            if (taintedTarget != TaintedTargetValue.Return)
+                            {
+                                IArgumentOperation argumentOperation = visitedArguments.FirstOrDefault(o => o.Parameter.Name == taintedTarget);
+                                Debug.Assert(argumentOperation != null, "Are the tainted data sources misconfigured?");
+                                this.CacheAbstractValue(argumentOperation, TaintedDataAbstractValue.CreateTainted(argumentOperation.Parameter, argumentOperation.Syntax, method));
+                            }
+                            else
+                            {
+                                result = TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                            }
                         }
+                    }
 
-                        return TaintedDataAbstractValue.NotTainted;
+                    if (this.DataFlowAnalysisContext.SourceInfos.IsSourcePasserMethod(
+                        method,
+                        visitedArguments,
+                        visitedArguments
+                            .Where(s => this.GetCachedAbstractValue(s).Kind == TaintedDataAbstractValueKind.Tainted)
+                            .Select(s => s.Parameter.Name)
+                            .ToImmutableArray(),
+                        out taintedParameterPairs))
+                    {
+                        foreach ((string source, string end) in taintedParameterPairs)
+                        {
+                            IArgumentOperation endArgumentOperation = visitedArguments.FirstOrDefault(o => o.Parameter.Name == end);
+                            Debug.Assert(endArgumentOperation != null, "Are the tainted data sources misconfigured?");
+                            SetTaintedForEntity(
```

Comment:
Also need an `if (endArgumentOperation == null)` so we can fail gracefully if it MethodMatcher isn't specific enough.